python/gstaichi/lang/_fast_caching/src_hasher.py

```diff
@@ -57,37 +60,45 @@ def store(cache_key: str, function_source_infos: Iterable[FunctionSourceInfo]) -
         - i.e. is our cache key still valid?
     - the python side cache contains information we will use to verify that our cache key is valid
         - ie the list of function source infos
+
+    Update! We are now going to store parameter pruning infomation, which is:
+    - used_py_dataclass_leaves: set[str]
     """
     if not cache_key:
         return
     cache = PythonSideCache()
     hashed_function_source_infos = function_hasher.hash_functions(function_source_infos)
-    cache_value_obj = CacheValue(hashed_function_source_infos=list(hashed_function_source_infos))
-    cache.store(cache_key, cache_value_obj.json())
+    cache_value_obj = CacheValue(
+        hashed_function_source_infos=list(hashed_function_source_infos),
+        used_py_dataclass_leaves=used_py_dataclass_leaves,
+    )
+    cache.store(cache_key, cache_value_obj.model_dump_json())
 
 
-def _try_load(cache_key: str) -> Sequence[HashedFunctionSourceInfo] | None:
+def _try_load(cache_key: str) -> tuple[Sequence[HashedFunctionSourceInfo], set[str]] | tuple[None, None]:
     cache = PythonSideCache()
     maybe_cache_value_json = cache.try_load(cache_key)
     if maybe_cache_value_json is None:
-        return None
+        return None, None
     try:
-        cache_value_obj = CacheValue.parse_raw(maybe_cache_value_json)
+        cache_value_obj = CacheValue.model_validate_json(maybe_cache_value_json)
     except (pydantic.ValidationError, json.JSONDecodeError, UnicodeDecodeError) as e:
         warnings.warn(f"Failed to parse cache file {e}")
-        return None
-    return cache_value_obj.hashed_function_source_infos
+        return None, None
+    return cache_value_obj.hashed_function_source_infos, cache_value_obj.used_py_dataclass_leaves
 
 
-def validate_cache_key(cache_key: str) -> bool:
+def load(cache_key: str) -> set[str] | None:
     """
     loads function source infos from cache, if available
     checks the hashes against the current source code
     """
-    maybe_hashed_function_source_infos = _try_load(cache_key)
-    if not maybe_hashed_function_source_infos:
-        return False
-    return function_hasher.validate_hashed_function_infos(maybe_hashed_function_source_infos)
+    maybe_hashed_function_source_infos, used_py_dataclass_leaves = _try_load(cache_key)
+    if maybe_hashed_function_source_infos is None or used_py_dataclass_leaves is None:
```

Comment:
Either your typing is wrong or the second clause of this condition is redundant.

Reply:
The current typing doesn't couple the two Nones together, though perhaps that is possible 🤔  Currently I have something like `tuple[Blah | None, Foo | None]`. But I suppose you are saying I could have osmthing like `tuple[Blah, Foo] | tuple[None, None]` ?

Reply:
Try replacing `Blah` and `Foo` by `int` in your example